csp_billing_adapter/archive.py

```diff
@@ -0,0 +1,33 @@
+#
+# Copyright 2024 SUSE LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Utility functions for handling a rolling dictionary archive."""
+
+
+def append_metering_records(
+    archive: list,
+    records: dict,
+    max_length: int
+):
+    """
+    Append metering records to the archive while maintaining max length.
+    """
+    archive.append(records)
```

Comment:
Should you check the length of the append vs current length before doing the append?

Reply:
Check the length of "records"? What would it be compared to or validated against?

Reply:
I see now, but I am not sure I understand what max_length is used for if you are returning [1:]

```
my_archive = []
for i in range(1,10):
    print(append_metering_records(my_archive, {"value": i}, 5))
```

output

```
[{'value': 1}]
[{'value': 1}, {'value': 2}]
[{'value': 1}, {'value': 2}, {'value': 3}]
[{'value': 1}, {'value': 2}, {'value': 3}, {'value': 4}]
[{'value': 1}, {'value': 2}, {'value': 3}, {'value': 4}, {'value': 5}]
[{'value': 2}, {'value': 3}, {'value': 4}, {'value': 5}, {'value': 6}]
[{'value': 2}, {'value': 3}, {'value': 4}, {'value': 5}, {'value': 6}, {'value': 7}]
[{'value': 2}, {'value': 3}, {'value': 4}, {'value': 5}, {'value': 6}, {'value': 7}, {'value': 8}]
[{'value': 2}, {'value': 3}, {'value': 4}, {'value': 5}, {'value': 6}, {'value': 7}, {'value': 8}, {'value': 9}]    
```

Reply:
One months worth of usage/records would be added at a time as a dictionary. So on any given call to `append_metering_records` there's never a chance for the archive to be more than 1 item too big. If it is the [1:] slice drops off the first (oldest) dictionary.

The test is missing the setting of the return value. E.g. 

```
my_archive = []
for i in range(1,10):
    my_archive = append_metering_records(my_archive, {"value": i}, 5 )
    print(my_archive)
```

Because it uses a slice it's not updating the list in a "pass by reference" style. The slice creates a new object.

Reply:
Thank you for explaining this.

Reply:
We could update the `append_metering_records` to do the update in that fashion. Either way works. Since it's mainly an internal function I don't think it matters too much which way. 🤷🏼